src/psycoptts/stratify_by_each_category_individually.py

```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+import numpy as np
+from pandas import DataFrame
+from typing import Optional, List, Tuple, Union, Dict
+import random
+
+
+def stratified_split_by_each_category(
+    df: DataFrame,
+    test_size=Union[float, int],
+    random_state=int,
+    stratify: Optional[List[str]] = None,
+) -> Tuple[DataFrame, DataFrame]:
+    no_cat_str = "no_cat"
+
+    # Generate target ns
+    target_n_in_test_by_cat = {
+        strat_col: df[df[strat_col] == 1].shape[0] * test_size for strat_col in stratify
+    }.copy()
+
+    if isinstance(test_size, float):
+        target_total_n_in_test = test_size * df.shape[0]
+
+    df["is_test"] = np.zeros(df.shape[0])
+
+    # Create a stack of idx to pull from
+    patient_ids_for_each_cat = {
+        col: df.index[df[col] == 1].tolist() for col in stratify
+    }
+
+    # Set random state for reproducibility
+    random.seed(random_state)
+
+    # Shuffle the idx
+    for key in patient_ids_for_each_cat.keys():
+        random.shuffle(patient_ids_for_each_cat[key])
+
+    # Mark idx that don't belong to any stratification-col
+    df[no_cat_str] = np.where(df[stratify].sum(axis=1) == 0, 1, 0)
+    patient_ids_for_each_cat[no_cat_str] = df.index[df[no_cat_str] == 1].tolist()
+
+    # Initialise counters for while-loop
+    n_in_test = 0
+    n_in_test_by_category = defaultdict(lambda: 0)
+    all_cats_reached_target_n_in_test = False
+
+    while n_in_test < target_total_n_in_test:
+        ratio_target_current_prop = {
+            cat: n_in_test_by_category[cat] / target_n_in_test_by_cat[cat]
+            for cat in stratify
+        }
+
+        if get_minimum_val_in_dict(ratio_target_current_prop) >= 1:
+            all_cats_reached_target_n_in_test = True
+
+        if all_cats_reached_target_n_in_test:
+            cat_to_add_to = no_cat_str
+        else:
+            # Sample from the category that is the furthest from the desired prop
+            cat_to_add_to = min(
+                ratio_target_current_prop, key=ratio_target_current_prop.get
+            )
+
+        idx = patient_ids_for_each_cat[cat_to_add_to].pop()
+
+        # If an idx belongs to multiple categories, also remove if from those that aren't currently cat_to_add_to
+        # This is __super__ inefficient (95% of current runtime), but running for 120.000 idxs takes 45 seconds.
+        # Acceptable runtime for current application.
+        remove_value_from_all_lists_it_exists_in(
+            value=idx, dict_of_lists=patient_ids_for_each_cat
+        )
```

Comment:
An easy fix here is to have the list be sets (reduced the lookup complexity from O(n) to O(1)).